File: puzzles/go/there-is-no-spoon1.go
package main

import "fmt"
import "os"
import "bufio"

func main() {
  const EMPTY = byte('.')
  const NODE = byte('0')
  var grid []string

  scanner := bufio.NewScanner(os.Stdin)
  scanner.Buffer(make([]byte, 1000000), 1000000)

  // width: the number of cells on the X axis
  var width int
  scanner.Scan()
  fmt.Sscan(scanner.Text(), &width)
  
  // height: the number of cells on the Y axis
  var height int
  scanner.Scan()
  fmt.Sscan(scanner.Text(), &height)
  
  for i := 0; i < height; i++ {
    scanner.Scan()
    line := scanner.Text()
    grid = append(grid, line)
  }
  
  for y := 0; y < height; y++ {
    for x := 0; x < width; x++ {
      if grid[y][x] == NODE {
        // node
        fmt.Printf("%v %v ", x, y)

        // right neighbor
        var neighbor = EMPTY
        for neighborX := x + 1; neighborX < width; neighborX++ {
          neighbor = grid[y][neighborX]
          if neighbor == NODE {
            fmt.Printf("%v %v ", neighborX, y)
            break
          }
        }

        if neighbor == EMPTY {
          fmt.Print("-1 -1 ")
        }

        // bottom neighbor
        neighbor = EMPTY
        for neighborY := y + 1; neighborY < height; neighborY++ {
          neighbor = grid[neighborY][x]
          if neighbor == NODE {
            fmt.Printf("%v %v ", x, neighborY)
            break
          }
        }

        if neighbor == EMPTY {
          fmt.Print("-1 -1 ")
        }

        fmt.Println()
      }
    }
  }
}

File: puzzles/cpp/there-is-no-spoon1.cpp
#include <iostream>
#include <vector>

using namespace std;

int main() {
  const char EMPTY = '.';
  const char NODE = '0';
  vector<string> grid;

  int width; // the number of cells on the X axis
  cin >> width; cin.ignore();
  int height; // the number of cells on the Y axis
  cin >> height; cin.ignore();
  for (int i = 0; i < height; i++) {
    string line;
    getline(cin, line);
    grid.push_back(line);
  }

  for (int y = 0; y < height; y++) {
    for (int x = 0; x < width; x++) {
      if (grid[y][x] == NODE) {
        // node
        cout << x << ' ' << y << ' ';

        // right neighbor
        char neighbor = EMPTY;
        for (int neighbor_x = x + 1; neighbor_x < width; neighbor_x++) {
          neighbor = grid[y][neighbor_x];
          if (neighbor == NODE) {
            cout << neighbor_x << ' ' << y << ' ';
            break;
          }
        }

        if (neighbor == EMPTY) {
          cout << "-1 -1 ";
        }

        // bottom neighbor
        neighbor = EMPTY;
        for (int neighbor_y = y + 1; neighbor_y < height; neighbor_y++) {
          neighbor = grid[neighbor_y][x];
          if (neighbor == NODE) {
            cout << x << ' ' << neighbor_y << ' ';
            break;
          }
        }

        if (neighbor == EMPTY) {
          cout << "-1 -1 ";
        }

        cout << endl;
      }
    }
  }
}
